modules/rtdModule/index.js

```diff
@@ -159,130 +180,147 @@ export function init(config) {
-  });
-  events.on(CONSTANTS.EVENTS.BEFORE_REQUEST_BIDS, (args) => {
-    subModules.forEach(sm => { sm.updateBidRequest && sm.updateBidRequest(args, sm.config) })
+    getAdUnitTargeting(args);
+    subModules.forEach(sm => { sm.onAuctionEndEvent && sm.onAuctionEndEvent(args, sm.config, _userConsent) })
   });
   events.on(CONSTANTS.EVENTS.BID_RESPONSE, (args) => {
-    subModules.forEach(sm => { sm.updateBidResponse && sm.updateBidResponse(args, sm.config) })
+    subModules.forEach(sm => { sm.onBidResponseEvent && sm.onBidResponseEvent(args, sm.config, _userConsent) })
   });
 }
 
 /**
- * get data from sub module
- * @param {AdUnit[]} adUnits received from auction
- * @param {function} callback callback function on data received
+ * loop through configured data providers If the data provider has registered getBidRequestData,
+ * call it, providing reqBidsConfigObj, consent data and module params
+ * this allows submodules to modify bidders
+ * @param {Object} reqBidsConfigObj required; This is the same param that's used in pbjs.requestBids.
+ * @param {function} fn required; The next function in the chain, used by hook.js
  */
-export function getProviderData(adUnits, callback) {
-  /**
-   * invoke callback if one of the conditions met:
-   * timeout reached
-   * all submodules answered
-   * all sub modules configured "waitForIt:true" answered (as long as there is at least one configured)
-   */
-
-  const waitForSubModulesLength = subModules.filter(sm => sm.config && sm.config.waitForIt).length;
-  let callbacksExpected = waitForSubModulesLength || subModules.length;
-  const shouldWaitForAllSubModules = waitForSubModulesLength === 0;
-  let dataReceived = {};
-  let processDone = false;
-  const dataWaitTimeout = setTimeout(done, _moduleConfig.auctionDelay || _moduleConfig.timeout || DEF_TIMEOUT);
-  subModules.forEach(sm => {
-    sm.getData(adUnits, onDataReceived.bind(sm));
+export function setBidRequestsData(fn, reqBidsConfigObj) {
+  initSubModules();
+
+  // delay bidding process only if auctionDelay > 0
+  if (!_moduleConfig.auctionDelay || _moduleConfig.auctionDelay < 1) {
+    return exitHook();
+  }
+
+  const relevantSubModules = subModules.filter(sm => typeof sm.getBidRequestData === 'function');
+  const prioritySubModules = relevantSubModules.filter(sm => !!(sm.config && sm.config.waitForIt));
+  let callbacksExpected = prioritySubModules.length;
+  let isDone = false;
+  let waitTimeout;
+
+  if (!relevantSubModules.length) {
+    return exitHook();
+  }
+
+  if (prioritySubModules.length) {
+    waitTimeout = setTimeout(exitHook, _moduleConfig.auctionDelay);
+  }
+
+  relevantSubModules.forEach(sm => {
+    sm.getBidRequestData(reqBidsConfigObj, onGetBidRequestDataCallback.bind(sm), sm.config, _userConsent)
   });
 
-  function onDataReceived(data) {
-    if (processDone) {
-      return
+  if (!prioritySubModules.length) {
+    return exitHook();
+  }
+
+  function onGetBidRequestDataCallback(auctionId) {
```

Comment:
Is the auctionId always available when this gets here? From looking at the code in src/prebid.js, not clear that it is. Maybe a change is needed there.

Reply:
We actually not doing anything with the auction ID, it's there by mistake (debugging leftovers), I'll remove it.

Reply:
Guess I don't understand how this code works when there are multiple concurrent auctions. This can and will happen.

Maybe we need `callbacksExpected` and `waitTimeout` to be a hashes indexed by auctionId?